util/placementpolicy/policy.go

```diff
@@ -15,23 +15,80 @@
 package placementpolicy
 
 import (
+	"fmt"
+	"strings"
+
 	"github.com/pingcap/parser/model"
 )
 
+// PlacementSettings is the settings of the placement
+type PlacementSettings struct {
+	PrimaryRegion       string `json:"primary_region"`
+	Regions             string `json:"regions"`
+	Learners            uint64 `json:"learners"`
+	Followers           uint64 `json:"followers"`
+	Voters              uint64 `json:"voters"`
+	Schedule            string `json:"schedule"`
+	Constraints         string `json:"constraints"`
+	LeaderConstraints   string `json:"leader_constraints"`
+	LearnerConstraints  string `json:"learner_constraints"`
+	FollowerConstraints string `json:"follower_constraints"`
+	VoterConstraints    string `json:"voter_constraints"`
+}
+
+func (p *PlacementSettings) String() string {
+	settings := make([]string, 0)
```

Comment:
Maybe `strings.builder`.

Reply:
`strings.Join` also uses string builder to join strings so I think the performance not varys too much. But I modify it is easy.

Reply:
That is wrong in some aspects. `strings.Builder` works like `bytes.Buffer`, where you only (re)alloc one region of memory. 

The problem of your code is on `[]string`, where you may allocated `n` strings. And `n` may be larger the number of allocations that will be done in  `strings.Builder`.